Feature/add post handler

morses-code opened the pull request:
### What

Added an initial attempt at the post handler for a session.

### How to review

run `make debug`
make a `post` request to `http://{host}:{port}/session`

body sample:
~~`{"id":"123","email":"me@me.com","start":"2020-04-22T07:00:00.000Z"}`~~
`{"email":"me@me.com"}`

### Who can review

Anyone

morses-code committed:
Added session struct

- Created session structs with json model
- added function for marshalling json model

morses-code committed:
Initial draft of CreateSessionHandler

- Added routing to the api
- Created the handler

morses-code committed:
Removed useless code

- Removed some code that was for visualisation, added nothing functionally

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,25 @@
+	"net/http"
+)
+
+func CreateNewSessionHandler() http.HandlerFunc {
```
Name is slightly confusing - is it creating a new session or creating a new session handler? I would probably go with `CreateSessionHandlerFunc` - as this function is not the handler but returns a http handler func. Nit picking but you asked for a review 😉

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,25 @@
+	"github.com/ONSdigital/log.go/log"
+	"net/http"
+)
+
```
Going put this now to save you an arse-ache at the end - add Godoc for your public methods/functions as you go along.

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,25 @@
+			http.Error(w, "Failed to decode session json", http.StatusBadRequest)
+			return
+		}
+
```
For the purposes of this task - add a log message here to spew out the session details.

daiLlew commented on session/session.go:
```diff
@@ -0,0 +1,34 @@
+	"time"
+)
+
+type Session struct {
```
Godoc on your public types.

daiLlew commented on session/session.go:
```diff
@@ -0,0 +1,34 @@
+	Start time.Time `json:"start"`
+}
+
+const (
```
Constants and vars should go before types, functions and methods.

daiLlew commented:
Minor comments but functionally is fine. Add some unit tests to prove this functionality - I find a TDD approach is helpful 
- Write a test for what you have.
- Add the next minimum amount of code.
- Repeat.

I personally work through all of the unhappy paths writing tests for them. Consider when happens if:
- The body is empty
- The body is invalid json
- The session is missing fields - email, id etc.

Probably best to save this until you have written the code but you will also need to update the Swagger spec to include this new endpoint. Again it will be easier to do this as write each endpoint instead of at the end.

morses-code committed:
Added tests and some PR comment addresses

- Removed the hello code
- Added handler tests
- Started to look at the swagger file
- Updated config with local port
- Added some godoc

morses-code committed:
Commented out failing test

- Commented out failing test for now.

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,28 @@
+		var sess session.Session
+		err := json.NewDecoder(r.Body).Decode(&sess)
+		if err != nil {
+			log.Event(ctx, "unable to decode session", log.Error(err), log.ERROR)
```
Nothing technically wrong with this statement but I think the team would generally use the term `unmarshall`. Very minor point but it makes this message clearer and keeps it consistent with the rest of our logging.

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,69 @@
+		})
+	})
+
+	Convey("Give a request to /session with body with all elements", t, func() {
```
_Given a valid create session request_?

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,69 @@
+			Email: "me@me.com",
+			Start: time.Now(),
+		}
+		sessJSON, _ := sess.MarshalJSON()
```
Unlikely to happen in the test but its better to do
```
sessJSON, err := sess.MarshalJSON()
So(err, ShouldBeNil)
```

morses-code replied:
Makes sense. Initially had no idea what to do with the error.

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,28 @@
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		var sess session.Session
```
Thinking about it I think the body of this request should probably be its own model. I'm not a fan of reusing models I find it can become complicated and messy - as you end up with optional fields and validation becomes dependant on the use case 🤮. Also posting only some of the session details to create a session feels a bit weird.

I think we should have a model for creating a new session:
```
type CreateSessionEntity struct {
   Email string    `json:"email"`
}
```
I think this is much clearer - you send a `CreateSessionEntity` body and if successful receive a `Session` back.

morses-code replied:
Would you be able to run through this with me?

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,28 @@
+		log.Event(ctx, "session created", log.Data{"session": sess}, log.INFO)
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusCreated)
```
This endpoint needs to return the session that was created/allow the caller to retrieve it.

I can see this being contentious within the team but adhering strictly to REST principals this should return a `Content-Location` header in the response with a URI to the created resource.

```
Content-Location: /session/{ID}
```
The caller can use this uri/ID to make a `GET: /session{ID)` request to get the session entity back if they require it.

However, most of our APIs in this case would just return the session body from the `POST` request - personally I prefer the first approach. This way feels like we are bastardising a well defined standard.

The caller does have to potentially make multiple requests but endpoints are well defined and **perform a single action**. Their responsibility is clearly defined and tests are easier to write. Additionally I think you end up with a much more flexible API as the actions are specific enough it enables the caller can chain them together in the way they desire to meet their need.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Location#Examples

morses-code committed:
Updated wording in handlers and updated tests

- Changed decode to unmarshal in handlers.go
- Updated wording of test
- Added check for error being nil

morses-code committed:
Changes to creating a session

- Added customer decoder that will return errors for missing fields
- New struct for request body for creating a new session
- New function for creating a new session
- Added content-location to post response
- Added new test
- Added errors.go

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,31 @@
+
+		var c CreateSessionEntity
+		if err := decoder.Decode(r, &c); err != nil {
+			log.Event(ctx, "missing a required field", log.Error(err), log.ERROR)
```
How do you know this is missing a field? If this fails it feels more like a 500 error - we failed to parse the request.

If the body is empty json I think this will still be successfully marshalled - it will just be an empty struct. You'll need to add a validation step after body has been marshalled into the struct. The validation check will allow you to be more specific in what HTTP status code should be returned.

morses-code replied:
Using the custom decoder, when passing an empty body it returns the empty field error.

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,65 @@
+			Email: "me@me.com",
+			Start: time.Now(),
+		}
+		sessJSON, err := sess.MarshalJSON()
```
```
So(err, ShouldBeNil)
```
Being thorough

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,65 @@
+		Convey("When the request is handled by the router", func() {
+			sessionHandler.ServeHTTP(resp, req)
+
+			Convey("Then the response should be 201", func() {
```
`Then the expected success response is returned`  assert that:

- The status is as expected
- The response `Content-Location` header exists and fits the expected format `/session{id}`

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,65 @@
+		})
+	})
+
+	Convey("Give a request to /session with a body with missing elements", t, func() {
```
`"Give a request create session request with no email address"`

daiLlew commented on decoder/decoder.go:
```diff
@@ -0,0 +1,18 @@
+		return err
+	}
+	return v.OK()
+}
```
Don't think this is needed - you still have to do `if err != nil` so you aren't saving any duplication.

morses-code replied:
Using this I am able to be sure that there is a required field missing from the JSON.

daiLlew replied:
I think thats coupling json deserialisation and validation. If the field is missing that a different error to being unable to unmarshall the JSON.

daiLlew commented on session/session.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// OK checks CreateSessionEntity is valid
+func (c *CreateSessionEntity) OK() error {
```
`OK` is quite generic. If you were new to this code and were scanning through it you wouldn't know what this does just by reading the name. Personal preference here but I think good names are really valuable, this is a bit longer but I think if you read this in the code that is using it its instantly more obvious what the function does.
```
session.Validate_<NAME_OF_THING>(x X)
```

(See comment above about renaming the model).

daiLlew commented on session/session.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// CreateSessionEntity is the structure of the request needed to create a session
+type CreateSessionEntity struct {
```
I'm beginning to dislike this name (which is my fault not yours). I think something along the lines of `NewSessionDetails` sounds nicer. `SREXDFUYGHIOJOJEntity` feels a bit too Enterprise Java.

daiLlew commented on session/session.go:
```diff
@@ -0,0 +1,69 @@
+func newID() string {
+	id, err := uuid.NewUUID()
+	if err != nil {
+		return ""
```
Log here. We would want to see this has gone wrong.

morses-code replied:
Need the context for the logging. How would I supply this from here?

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,31 @@
+			return
+		}
+
+		sess := CreateNewSession(c.Email)
```
As discussed:

Create a new interface type defined here in the handler package which will be responsible for the sessions logic
```
type SessionService interface {
    CreateNewSession(email string) session.Session {
}
```
We'll add more methods to this interface as we flesh out the API. For now we're adding the only method we need right now. Update the create handler function to take an argument of type `SessionService`
```
CreateSessionHandlerFunc(sessionService SessionService)
```

Use the session service to create the new session when handling a request
```
sessionService.CreateNewSession(c.Email)
```
This decouples the `handler` and `session` packages and gives you more control. You can now provided the implementation you want use when you create the handler. For the app this will be a real implementation for the test you will use a mock.

daiLlew commented on session/session.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// CreateNewSession creates a new session using email parameter
+func CreateNewSession(email string) Session {
```
Create a new service struct type in this package
```
type Service struct {
   // whatever fields are needed
}
```

Update the `CreateNewSession` func adding a receiver for the service struct - this makes it a method of the service struct instead of a package function.
```
func (s *Service) CreateNewSession(email string) Session {}
```
This struct now implements the interface you defined in the handler package so when you create the session handler you can provide an instance of this type as the session service.

morses-code committed:
Addressed PR comments, some renaming 

- Added back full package name in handlers
- Changed naming in the decoder
- Updated naming of CreateSessionEntity
- Added logging for possible NewUUID() error

morses-code committed:
Various tidying up, removing redundant stuff. updating tests

- Removed apierrors
- Removed decoder
- Generated mocksessions
- Updated handlers, decoupled json decoding and validation
- Added nop
- Tests now using mocks

daiLlew commented on api/handlers.go:
```diff
@@ -0,0 +1,49 @@
+	"net/http"
+)
+
+type Sessions interface {
```
Godoc for public types methods and functions.

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,139 @@
+
+	Convey("Given a request to /session with a body with missing elements", t, func() {
+		mockSessions := &SessionsMock{
+			NewFunc: func(email string) (*session.Session, error) {
+				return &session.Session{
+					ID:    "1234",
+					Email: "",
+				}, nil
+			},
```
This isn't needed for this test case - it can be deleted.

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,139 @@
+		})
+	})
+
+	Convey("Given a valid create session request but UUID fails to generate", t, func() {
```
This is a potential error however this exposing details of the inner workings of the session package - we don't want to expose that and couple them together.

For this test we are testing what happens when `sessions.New()` returns an error - we don't need to know the details of the error we only care that the handle responds as expected when this happens.

Your test code is fine but change the context to something like 
**"Given a create new session returns an error"**

daiLlew commented on api/handlers_test.go:
```diff
@@ -0,0 +1,139 @@
+		}
+		sessionHandler := CreateSessionHandlerFunc(mockSessions)
+
+		sess := session.NewSessionDetails{
+			Email: "test@test.com",
+		}
+		sessJSON, err := json.Marshal(sess)
+		So(err, ShouldBeNil)
```
We've repeated this code 3+ times now which I think means you could probably pull this out into a helper func. A general rule of thumb I follow is - if you duplicate it 3 times consider abstracting it.

nshumoogum commented on api/handlers_test.go:
```diff
@@ -0,0 +1,139 @@
+			sessionHandler.ServeHTTP(resp, req)
+
+			Convey("Then return an error response", func() {
+				So(resp.Code, ShouldEqual, http.StatusInternalServerError)
```
Should this return bad request - 400? Internal server error would mean something went wrong with the sessions API

nshumoogum commented on api/handlers.go:
```diff
@@ -0,0 +1,49 @@
+		var c session.NewSessionDetails
+		err := json.NewDecoder(r.Body).Decode(&c)
+		if err != nil {
+			log.Event(ctx, "failed to unmarshal", log.Error(err), log.ERROR)
```
Maybe update log and http error to something like `failed to decode request body`

nshumoogum commented on api/handlers.go:
```diff
@@ -0,0 +1,49 @@
+		err := json.NewDecoder(r.Body).Decode(&c)
+		if err != nil {
+			log.Event(ctx, "failed to unmarshal", log.Error(err), log.ERROR)
+			http.Error(w, "Failed to unmarshal", http.StatusInternalServerError)
```
Im not sure this is an internal server error, if they don't send a body then it is a bad request 400
```
err := json.NewDecoder(r.Body).Decode(&c)
switch {
case err == io.EOF:
    // empty body
case err != nil:
    // other error
}

or

if err := json.NewDecoder(r.Body).Decode(&c); err != nil {
    if err == io.EOF {
        // empty body
    } else {
       // other error
    }
}
```

nshumoogum replied:
Arguably if they sent in bad json e.g. `{"email": "testuser@hotmail.com"` this I believe will fall into other error and again is a client error and we should be informing them they have sent a bad request 400. 500 internal server error is based on an issue server side and that the client should be able to retry their request and possibly get a different status code (2**)

So should an error falling out of the decode method ever result in a response status of 500 instead of 400? I personally don't think so and checking the package I can't see where it would be an issue server side.

nshumoogum commented on api/handlers_test.go:
```diff
@@ -0,0 +1,139 @@
+			sessionHandler.ServeHTTP(resp, req)
+
+			Convey("Then return an error response", func() {
+				So(resp.Code, ShouldEqual, http.StatusInternalServerError)
```
If it is a bad request it should return 400 - http.StatusBadRequest

nshumoogum commented on session/session.go:
```diff
@@ -0,0 +1,63 @@
+func newID() string {
+	id, err := uuid.NewUUID()
+	if err != nil {
+		log.Event(context.Background() /*TODO - setting the context?*/, "unable to generate id", log.Error(err), log.ERROR)
```
I would pass in context to newID instead of instantiating a new context, this way we can follow the request being made to the failure in this method:

```
func newID(ctx context.Context) string {
   ...
}
```

This will force you to update your interface and regenerate your mocks and lastly update the tests, with the tests I would instantiate a global testContext value for use, like so:

`var testContext = context.Background()`

daiLlew replied:
Generally you should either handle an error **or** return it and let the caller handle it. In this instance I don't think we need to log the error and can just return it - you'll need to update the function signature but this also means you don't need to pass in the `context` as a parameter.

nshumoogum commented on swagger.yaml:
```diff
@@ -10,33 +10,26 @@ basePath: "/v1"
+          description: Created
           schema:
-            $ref: "#/definitions/HelloResponse"
-        500:
```
May need to reintroduce this as I see internal error could be generated through uuid method when calling POST /sessions

nshumoogum commented on swagger.yaml:
```diff
@@ -10,33 +10,26 @@ basePath: "/v1"
+        400:
+          description: Bad Request
+          schema:
+            $ref: "#/responses/BadRequest"
```
Missing definition of BadRequest in responses object - you could remove this as you have added a description directly on the status code 400

The reason for moving it to responses object and then referencing it is so it can be re-used by other parts of the spec, e.g. used on another endpoint

daiLlew replied:
You also need to add the type for the request body.

nshumoogum commented on swagger.yaml:
```diff
@@ -10,33 +10,26 @@ basePath: "/v1"
-            $ref: "#/definitions/HelloResponse"
-        500:
-          $ref: '#/responses/InternalError'
+            $ref: "#/definitions/SessionResponse"
```
SessionResponse missing - you need to define the json response body

nshumoogum commented on swagger.yaml:
```diff
@@ -10,33 +10,26 @@ basePath: "/v1"
-    type: object
-    properties:
-      message:
-        type: string
```
Regarding the swagger spec - I would copy the raw file of swagger.yaml and paste it into https://editor.swagger.io/ (left hand side of web page) this will inform you of any errors in your swagger spec and also has a helpful ui for reading the spec (right hand side of web browser)

morses-code committed:
Response updates, test updates, Session updates

- Added godoc for Sessions interface
- Updated working of error message for unmarshal error
- Updated response codes in tests to return 400
- Removed redundant set up for mockSessions
- Added error returns for CreateNewSession and newID

morses-code committed:
Tried to update swagger.yaml

- Made changes to update swagger.yaml

daiLlew committed:
fix up swagger spec

nshumoogum commented on config/config.go:
```diff
@@ -23,7 +23,7 @@ func Get() (*Config, error) {
-		BindAddr:                   ":",
+		BindAddr:                   ":24400",
 		GracefulShutdownTimeout:    5 * time.Second,
 		HealthCheckInterval:        10 * time.Second,
```
I know you inherited this code but do you mind updating this configuration, default HealthCheckInterval should be 30seconds and HealthcheckCriticalTimeout should be 90 seconds

nshumoogum replied:
Bump

```
                HealthCheckInterval:        30 * time.Second,
                HealthCheckCriticalTimeout: 90 * time.Second,
```

morses-code committed:
Updated healthcheck config, added helper function in tests

morses-code committed:
Updated test value
